File: go-backend-interview-prep/06_concurrency/06_timeouts_and_timers/main.py
"""Timeouts and timers -- Python equivalent of the Go example.

Python uses asyncio.wait_for for timeouts and asyncio.sleep for timers.
There is no channel-based timer; async sleep is the closest analogy.
"""

import asyncio


async def main() -> None:
    # --- Example 1: timeout with asyncio.wait_for ---
    print("=== asyncio.wait_for timeout ===")

    async def slow_work():
        await asyncio.sleep(0.2)
        return "result"

    try:
        v = await asyncio.wait_for(slow_work(), timeout=0.1)
        print(f"  received: {v}")
    except asyncio.TimeoutError:
        print("  timed out after 100ms")

    # --- Example 2: one-shot timer (asyncio.sleep) ---
    print("\n=== One-shot timer ===")
    await asyncio.sleep(0.1)
    print("  timer fired after 100ms")

    await asyncio.sleep(0.05)
    print("  timer fired again after 50ms (reset)")

    # --- Example 3: periodic ticker (loop + sleep) ---
    print("\n=== Periodic ticker ===")
    count = 0
    interval = 0.08
    deadline = asyncio.get_event_loop().time() + 0.35

    while True:
        await asyncio.sleep(interval)
        now = asyncio.get_event_loop().time()
        if now >= deadline:
            print(f"  stopped after {count} ticks")
            break
        count += 1
        print(f"  tick {count}")


if __name__ == "__main__":
    asyncio.run(main())

File: go-backend-interview-prep/06_concurrency/06_timeouts_and_timers/main.go
package main

import (
	"fmt"
	"time"
)

func main() {
	// --- Example 1: time.After timeout ---
	fmt.Println("=== time.After timeout ===")
	ch := make(chan string)

	go func() {
		time.Sleep(200 * time.Millisecond)
		ch <- "result"
	}()

	select {
	case v := <-ch:
		fmt.Println("  received:", v)
	case <-time.After(100 * time.Millisecond):
		fmt.Println("  timed out after 100ms")
	}

	// --- Example 2: time.NewTimer (reusable) ---
	fmt.Println("\n=== NewTimer ===")
	timer := time.NewTimer(100 * time.Millisecond)

	<-timer.C
	fmt.Println("  timer fired after 100ms")

	// Reset and use again
	timer.Reset(50 * time.Millisecond)
	<-timer.C
	fmt.Println("  timer fired again after 50ms (reset)")

	// --- Example 3: time.NewTicker (periodic) ---
	fmt.Println("\n=== NewTicker (periodic) ===")
	ticker := time.NewTicker(80 * time.Millisecond)
	defer ticker.Stop()

	done := time.After(350 * time.Millisecond)
	count := 0

	for {
		select {
		case t := <-ticker.C:
			count++
			fmt.Printf("  tick %d at %v\n", count, t.Format("15:04:05.000"))
		case <-done:
			fmt.Printf("  stopped after %d ticks\n", count)
			return
		}
	}
}
